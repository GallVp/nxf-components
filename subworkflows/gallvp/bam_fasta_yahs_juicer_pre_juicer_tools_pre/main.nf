include { SAMTOOLS_FAIDX                    } from '../../../modules/gallvp/samtools/faidx/main'
include { JUICEBOXSCRIPTS_MAKEAGPFROMFASTA  } from '../../../modules/gallvp/juiceboxscripts/makeagpfromfasta/main'
include { JUICEBOXSCRIPTS_AGP2ASSEMBLY      } from '../../../modules/gallvp/juiceboxscripts/agp2assembly/main'
include { CUSTOM_ASSEMBLY2BEDPE             } from '../../../modules/gallvp/custom/assembly2bedpe/main'
include { YAHS_JUICERPRE                    } from '../../../modules/gallvp/yahs/juicerpre/main'
include { SORT                              } from '../../../modules/gallvp/sort/main'
include { JUICERTOOLS_PRE                   } from '../../../modules/gallvp/juicertools/pre/main'

workflow BAM_FASTA_YAHS_JUICER_PRE_JUICER_TOOLS_PRE {

    take:
    ch_bam                          // channel: [ val(meta), bam ]; meta: [ id, ref_id ]
    ch_fasta                        // channel: [ val(meta2), fasta ]; meta2: [ id ]
                                    // meta2.id = meta.ref_id
    val_assembly_mode               // true|false; Turn on/off assembly mode '-a' for YAHS_JUICERPRE

    main:
    ch_versions                     = Channel.empty()

    // MODULE: SAMTOOLS_FAIDX
    SAMTOOLS_FAIDX ( ch_fasta,
        [ [] , []],
        true // get_sizes
    )

    ch_fasta_fai_sizes              = ch_fasta
                                    | join(SAMTOOLS_FAIDX.out.fai)
                                    | join(SAMTOOLS_FAIDX.out.sizes)
    ch_versions                     = ch_versions.mix(SAMTOOLS_FAIDX.out.versions.first())

    // MODULE: JUICEBOXSCRIPTS_MAKEAGPFROMFASTA
    JUICEBOXSCRIPTS_MAKEAGPFROMFASTA ( ch_fasta )

    ch_fasta_fai_sizes_agp          = ch_fasta_fai_sizes
                                    | join(JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.agp)
    ch_versions                     = ch_versions.mix(JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.versions.first())

    // MODULES: JUICEBOXSCRIPTS_AGP2ASSEMBLY > CUSTOM_ASSEMBLY2BEDPE
    JUICEBOXSCRIPTS_AGP2ASSEMBLY (
        val_assembly_mode
        ? JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.agp
        : Channel.empty()
    )

    CUSTOM_ASSEMBLY2BEDPE (
        JUICEBOXSCRIPTS_AGP2ASSEMBLY.out.assembly
    )

    ch_versions                     = ch_versions
                                    | mix(JUICEBOXSCRIPTS_AGP2ASSEMBLY.out.versions.first())
                                    | mix(CUSTOM_ASSEMBLY2BEDPE.out.versions.first())

    // MODULE: YAHS_JUICERPRE
    ch_yahs_juicerpre_inputs        = ch_bam
                                    | map { meta, bam ->
                                        [ meta.ref_id, meta, bam ]
                                    }
                                    | join(
                                        ch_fasta_fai_sizes_agp
                                        | map { meta2, fasta, fai, sizes, agp ->
                                            [ meta2.id, fasta, fai, sizes, agp ]
                                        }
                                    )
                                    | multiMap { _ref_id, meta, bam, _fasta, fai, sizes, agp ->
                                        bam: [ meta, bam ]
                                        agp: agp
                                        fai: fai
                                        sizes: [ meta, sizes ]
                                    }

    YAHS_JUICERPRE (
        ch_yahs_juicerpre_inputs.bam,
        ch_yahs_juicerpre_inputs.agp,
        ch_yahs_juicerpre_inputs.fai
    )

    ch_versions                     = ch_versions.mix(YAHS_JUICERPRE.out.versions.first())

    // MODULE: SORT
    ch_sort_input                   = ! val_assembly_mode
                                    ? YAHS_JUICERPRE.out.txt
                                    : Channel.empty()
    SORT ( ch_sort_input )
    ch_versions                     = ch_versions.mix(SORT.out.versions.first())

    // MODULE: JUICERTOOLS_PRE
    ch_juicertools_pre_inputs       = (
                                        val_assembly_mode
                                        ? YAHS_JUICERPRE.out.txt
                                        : SORT.out.sorted
                                    )
                                    | join (
                                        val_assembly_mode
                                        ? YAHS_JUICERPRE.out.sizes
                                        : ch_yahs_juicerpre_inputs.sizes
                                    )
                                    | multiMap { meta, sorted, sizes ->
                                        sorted: [ meta, sorted ]
                                        sizes: sizes
                                    }

    JUICERTOOLS_PRE (
        ch_juicertools_pre_inputs.sorted,
        ch_juicertools_pre_inputs.sizes
    )

    ch_versions                     = ch_versions.mix(JUICERTOOLS_PRE.out.versions.first())

    emit:
    hic                             = JUICERTOOLS_PRE.out.hic                   // channel: [ meta, hic ]
    assembly                        = JUICEBOXSCRIPTS_AGP2ASSEMBLY.out.assembly // channel: [ meta, assembly ]
    bedpe                           = CUSTOM_ASSEMBLY2BEDPE.out.bedpe           // channel: [ meta, bedpe ]
    bed                             = CUSTOM_ASSEMBLY2BEDPE.out.bed             // channel: [ meta, bed ]
    versions                        = ch_versions                               // channel: [ versions.yml ]
}
