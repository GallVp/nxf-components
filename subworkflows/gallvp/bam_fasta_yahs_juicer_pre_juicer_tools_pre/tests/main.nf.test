nextflow_workflow {

    name "Test Subworkflow BAM_FASTA_YAHS_JUICER_PRE_JUICER_TOOLS_PRE"
    script "../main.nf"
    config "./nextflow.config"
    workflow "BAM_FASTA_YAHS_JUICER_PRE_JUICER_TOOLS_PRE"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/bam_fasta_yahs_juicer_pre_juicer_tools_pre"
    tag "samtools"
    tag "samtools/faidx"
    tag "yahs/juicerpre"
    tag "gunzip"
    tag "juiceboxscripts/makeagpfromfasta"
    tag "juiceboxscripts/agp2assembly"
    tag "custom/assembly2bedpe"
    tag "juicertools/pre"
    tag "sort"
    tag "juicer/indexbychr"

    setup {
        run("GUNZIP") {
            script "../../../../modules/gallvp/gunzip/main.nf"

            process {
                """
                input[0] = Channel.of([
                    [ id:'genome' ],
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.fasta.gz', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("hongyang - bam - fasta") {

        when {
            params {
                sort = '-k2,2d -k6,6d'
                juicerpre = ''
                juicertoolspre = ''
                assembly2bedpe = ''
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', ref_id:'genome' ],
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.bam', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = false // val_assembly_mode
                input[3] = false // val_use_index
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("hongyang - bam - fasta - assembly") {

        when {
            params {
                sort = '-k2,2d -k6,6d'
                juicerpre = '-a'
                juicertoolspre = '-r 1000'
                assembly2bedpe = '-s'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', ref_id:'genome' ],
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.bam', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = true // val_assembly_mode
                input[3] = false // val_use_index
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
