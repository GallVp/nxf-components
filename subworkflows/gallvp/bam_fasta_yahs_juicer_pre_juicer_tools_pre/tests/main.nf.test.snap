{
    "hongyang - bam - fasta - assembly": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.hic:md5,4bcdf548b354e625c81bb23061e365db"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.assembly:md5,48f07547635ccfd80eddc5b9a0d8040f"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.bedpe:md5,511c63569e96a71cc3c14d13c7198f15"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.bed:md5,ff0dc82bcd1f6e19089f1ab8a1a35a16"
                    ]
                ],
                "4": [
                    "versions.yml:md5,3d43ea23d3e662d51066c58979db7731",
                    "versions.yml:md5,4711b00bd24130a52a938ff8e4b47175",
                    "versions.yml:md5,8c78cd6cf3c2091662f765ced185c8e9",
                    "versions.yml:md5,ae6033bef2d56e168d019ac38ed69b94",
                    "versions.yml:md5,cdb6d83bee1c2f4c6142fa46e523ce8a",
                    "versions.yml:md5,d774719f806b09fd573710d9c356f726"
                ],
                "assembly": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.assembly:md5,48f07547635ccfd80eddc5b9a0d8040f"
                    ]
                ],
                "bed": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.bed:md5,ff0dc82bcd1f6e19089f1ab8a1a35a16"
                    ]
                ],
                "bedpe": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.bedpe:md5,511c63569e96a71cc3c14d13c7198f15"
                    ]
                ],
                "hic": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.hic:md5,4bcdf548b354e625c81bb23061e365db"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3d43ea23d3e662d51066c58979db7731",
                    "versions.yml:md5,4711b00bd24130a52a938ff8e4b47175",
                    "versions.yml:md5,8c78cd6cf3c2091662f765ced185c8e9",
                    "versions.yml:md5,ae6033bef2d56e168d019ac38ed69b94",
                    "versions.yml:md5,cdb6d83bee1c2f4c6142fa46e523ce8a",
                    "versions.yml:md5,d774719f806b09fd573710d9c356f726"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-21T16:42:49.981416"
    },
    "hongyang - bam - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.hic:md5,80d6f191efdceb8e7f7bbdcb5c522043"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,280e60b94741058316edef181421ce5b",
                    "versions.yml:md5,3d43ea23d3e662d51066c58979db7731",
                    "versions.yml:md5,4711b00bd24130a52a938ff8e4b47175",
                    "versions.yml:md5,ae6033bef2d56e168d019ac38ed69b94",
                    "versions.yml:md5,cdb6d83bee1c2f4c6142fa46e523ce8a"
                ],
                "assembly": [
                    
                ],
                "bed": [
                    
                ],
                "bedpe": [
                    
                ],
                "hic": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.hic:md5,80d6f191efdceb8e7f7bbdcb5c522043"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,280e60b94741058316edef181421ce5b",
                    "versions.yml:md5,3d43ea23d3e662d51066c58979db7731",
                    "versions.yml:md5,4711b00bd24130a52a938ff8e4b47175",
                    "versions.yml:md5,ae6033bef2d56e168d019ac38ed69b94",
                    "versions.yml:md5,cdb6d83bee1c2f4c6142fa46e523ce8a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-21T16:36:10.137402"
    }
}